File: src/finding_pairs_with_a_certain_sum_1865/solution.go
package finding_pairs_with_a_certain_sum_1865

import (
	"fmt"
	"sort"
)

type FindSumPairs struct {
	nums1        []int
	nums2        []int
	nums2ToCount map[int]int
}

func Constructor(nums1 []int, nums2 []int) FindSumPairs {
	sort.Slice(nums1, func(i, j int) bool {
		return nums1[i] < nums1[j]
	})
	nums2ToCount := make(map[int]int)

	for _, num2 := range nums2 {
		nums2ToCount[num2]++
	}

	return FindSumPairs{
		nums1:        nums1,
		nums2:        nums2,
		nums2ToCount: nums2ToCount,
	}
}

func (this *FindSumPairs) Add(index int, val int) {
	this.nums2ToCount[this.nums2[index]]--
	this.nums2[index] += val
	this.nums2ToCount[this.nums2[index]]++
}

func (this *FindSumPairs) Count(tot int) int {
	result := 0
	for _, num1 := range this.nums1 {
		if num1 >= tot {
			break
		}

		result += this.nums2ToCount[tot-num1]
	}

	return result
}

// Test
/**
 * Your FindSumPairs object will be instantiated and called as such:
 * obj := Constructor(nums1, nums2);
 * obj.Add(index,val);
 * param_2 := obj.Count(tot);
 */
func Test() {
	obj := Constructor([]int{1, 1, 2, 2, 2, 3}, []int{1, 4, 5, 2, 5, 4})
	fmt.Println(obj.Count(7))
	obj.Add(3, 2)
	fmt.Println(obj.Count(8))
	fmt.Println(obj.Count(4))
	obj.Add(0, 1)
	obj.Add(1, 1)
	fmt.Println(obj.Count(7))
}

File: src/finding_pairs_with_a_certain_sum_1865/solution.ts
class FindSumPairs {
  private readonly nums1: number[];
  private readonly nums2: number[];
  private nums2ToCount: Map<number, number>;

  constructor(nums1: number[], nums2: number[]) {
    this.nums1 = [...nums1].sort((a, b) => a - b);
    this.nums2 = nums2;
    this.nums2ToCount = new Map();

    for (const num of nums2) {
      this.nums2ToCount.set(num, (this.nums2ToCount.get(num) || 0) + 1);
    }
  }

  add(index: number, val: number): void {
    if (index < 0 || index >= this.nums2.length) {
      throw new Error("Index out of bounds");
    }

    const oldValue = this.nums2ToCount.get(this.nums2[index]);
    if (oldValue !== undefined) {
      this.nums2ToCount.set(this.nums2[index], oldValue - 1);
    }
    this.nums2[index] += val;
    this.nums2ToCount.set(this.nums2[index], (this.nums2ToCount.get(this.nums2[index]) || 0) + 1);
  }

  count(tot: number): number {
    let result = 0;

    for (const num1 of this.nums1) {
      if (num1 >= tot) {
        break;
      }

      result += this.nums2ToCount.get(tot - num1) || 0;
    }

    return result;
  }
}

export function test(): void {
  const obj = new FindSumPairs([1, 1, 2, 2, 2, 3], [1, 4, 5, 2, 5, 4]);
  console.log(obj.count(7));
  obj.add(3, 2);
  console.log(obj.count(8));
  console.log(obj.count(4));
  obj.add(0, 1);
  obj.add(1, 1);
  console.log(obj.count(7));
}
